v1.1 Options page

Pheil opened the pull request:
Added options panel, updated Seikichis tiff.min.js, minor code cleanup, fixed readme for changes and added note.

Pheil committed:
v1.1

Added options panel, updated Seikichis tiff.min.js,

d-Pixie commented on options/options.js:
```diff
@@ -0,0 +1,48 @@
+        } else {
+            MB = "32";
+        }
+        console.log(MB);
```
Maybe drop this line or put it behind a checkbox for debug output in the options panel? I assume you just missed it :)

d-Pixie commented on background.js:
```diff
@@ -1,5 +1,12 @@
+    if (obj.MB === undefined) {
+        mem_bytes = 32*1E6;
+    } else {
+        mem_bytes = obj.MB*1E6 ;
```
Implicit typecast here, since it is actually stored as `"32"` in persistent storage and not `32`. I would, ideally, want it to be saved as `32` instead, since it is actually a number and not a string we are interested in.

d-Pixie commented on options/options.js:
```diff
@@ -0,0 +1,48 @@
+// Saves options to chrome.storage.sync.
+function save_options() {
+  var MB = document.getElementById('MB').value;
```
I would like to see a better name for this. How about `"MaximumDecoderMemoryInMegabytes"` or `"DecoderMemoryLimitInMegabytes"` instead? For both the key and local variable.

d-Pixie replied:
A cast to integer is also needed here to ensure that we actually save a number and not a string, as per the first comment.

d-Pixie commented on options/options.js:
```diff
@@ -0,0 +1,48 @@
+        if(items.MB !== undefined) {
+            MB = items.MB;
+        } else {
+            MB = "32";
```
Can we declare this default in a constant at the top of the file instead? And it, too, should really be a number instead of a string.

d-Pixie commented on options/options.js:
```diff
@@ -0,0 +1,48 @@
+// stored in chrome.storage.
+function reset_options() {
+  // Use default value makeActive = true and displayBOM = true.
+    document.getElementById('MB').value = "32";
```
Use of default value.

d-Pixie replied:
Also, if the value stored in local storage is `64` at this point and you entered `"128"` in the textbox and pressed reset it would reset to `"32"` instead of `"64"` ... This should be the currently saved value, not the hard default...

Pheil replied:
I believe the point of the "Reset" button from the examples was to restore the defaults, not undo the last change.  Maybe re-name the button to "Restore"?

d-Pixie commented on options/options.js:
```diff
@@ -0,0 +1,48 @@
+    document.getElementById('MB').value = "32";
+    // Update status to let user know options were saved.
+    var status = document.getElementById('status');
+    status.textContent = 'Options reset, need to save.';
```
Nice touch.

d-Pixie commented on options/options.js:
```diff
@@ -0,0 +1,48 @@
+    chrome.storage.local.get('MB', function(items) {
+        var MB;
+        if(items.MB !== undefined) {
+            MB = items.MB;
```
If I remember correctly from the doc we also have to align the memory value to powers of 2 (see: https://groups.google.com/d/msg/emscripten-discuss/FJ9Y4qNg2to/6jqeFmZUeTAJ). Either round up to the nearest power of two if the user tries to set anything that isn't one or change the textfield to a dropdown?

d-Pixie reviewed (changes requested):
Thanks a lot for submitting this PR. I have some comments I would like you to address before we merge. I hope you see why I requested the changes, if not feel free to argue your case :)

Pheil committed:
v1.1 Update

Changes requested by @d-Pixie, added option to toggle debug output,
removed null from request.send() since it was throwing error in chrome.

d-Pixie commented on background.js:
```diff
@@ -1,126 +1,133 @@
-  request.send( null );
-  console.log( 'Finished XHR request' );
+    // Disable console.log if DebugOutput is false
+    if (obj.DebugOutput === undefined) {
```
Did you consider just doing:

    if( options.DebugOutput !== true ){
      console.log = function(){};
    }

Right now it would be on for anything that isn't `undefined` or `false`.

Pheil replied:
Oh, that makes more sense to do it that way.

d-Pixie commented on options/options.html:
```diff
@@ -6,13 +6,28 @@
+  <label>
+    <BR>
+    <!-- Must restart due to loading of background.js only once -->
+    <input type="checkbox" id="DebugOutput" title="Browser must be restarted for this to take effect.">
```
We can do that for them by reloading the extension: https://developer.chrome.com/extensions/runtime#method-reload

Pheil replied:
Sweet, didn't know about this method.  Works in Firefox too!

Pheil replied:
I'm assuming it has to do with the way Chrome reloads "unpacked extensions", but when testing Chrome resets ShowDebugOutput after reload.  It works "fine" in Firefox except for an apparent bug that throws ["ReferenceError: chrome is not defined[Learn More]"](https://bugzilla.mozilla.org/show_bug.cgi?id=1349558) so the add-ons manager needs manually refreshed.

d-Pixie replied:
Strange ... If the value is persisted in local storage and read back when the extension starts it should work. Are you sure you are reading it in when background.js is processed the first time? I guess that is when you'd have to set it ...

Pheil replied:
Aha, chrome.runtime.reload() was clearing the changed setting in Chrome but not Firefox.  Moved it to after I save the changed setting.

d-Pixie commented on options/options.html:
```diff
@@ -6,13 +6,28 @@
-  <button id="save">Save</button>
-  <button id="reset">Reset</button>
+  <button id="save" title="Save">Save</button>
+  <button id="reset" title="Restore default TIFF viewer options">Default</button>
```
Better :)

d-Pixie commented on options/options.js:
```diff
@@ -1,9 +1,16 @@
 // Saves options to chrome.storage.sync.
 function save_options() {
-  var MB = document.getElementById('MB').value;
+  var SaveDecoderMemoryLimitInMegabytes = parseInt(document.getElementById('DecoderMemoryLimitInMegabytes').value);
```
Great, a lot better now.

d-Pixie commented on options/options.js:
```diff
@@ -1,9 +1,16 @@
+"use strict";
+// Defaults
+var DecoderMemoryLimitInMegabytes = 32;
```
Use `const` instead of `var` for these and suffix both names with `Default` to clearly show later in the code that it is the default value we are using.

d-Pixie commented on options/options.js:
```diff
@@ -1,9 +1,16 @@
+"use strict";
+// Defaults
+var DecoderMemoryLimitInMegabytes = 32;
+var DebugOutput = false;
```
This should probably be `AllowDebugOutput` or `DebugOutputEnabled` or some such to better indicate that it is the option value and not, for example, a string of current debug output :)

d-Pixie commented on options/options.js:
```diff
@@ -16,22 +23,27 @@ function save_options() {
-        var MB;
-        if(items.MB !== undefined) {
-            MB = items.MB;
+    chrome.storage.local.get(['DecoderMemoryLimitInMegabytes', 'DebugOutput'], function(items) {
```
I would like it if we could consistently use `function( options ){ .. }` for these instead. Above I think you have `function( obj ){ .. }` for example and here it is `items` ...

d-Pixie commented on options/options.js:
```diff
@@ -16,22 +23,27 @@ function save_options() {
+            document.getElementById('DebugOutput').checked = items.DebugOutput;
         } else {
-            MB = "32";
+            document.getElementById('DebugOutput').checked = DebugOutput;
```
Here is a good example on the naming issue. It is not clear what the `DebugOutput` value you are using is coming from right now, looking at this line alone. Using a name like `DebugOutputEnabledDefault` will indicate clearly that it is a default value instead.

d-Pixie reviewed (commented):
Great improvements! I hope you don't mind me being picky with the naming :) I find that clear naming of concepts and variables help maintainability a lot in the long run. Thank you so much for your effort so far.

Pheil committed:
v1.1 Update 2

Added reload of extension after debug option changed, minor tweaks

d-Pixie commented:
Fantastic work! Thank you for the help, and patience :) Does this run on Firefox as well? If so I'll package it and put it in the extension/addon store for both Firefox and Chrome.

Pheil commented:
Thanks for taking the time to review and provide feedback.  Definitely got some good pointers to adjust my extensions.  Yes, this does work in Firefox.
